caret_vars.inc

```diff
@@ -0,0 +1,742 @@
+case 1:
+    switch ((U8)name[0]) {
+    case 1:
+        /* $^A */
+        /* FALLTHROUGH */;
+    case 3:
```

Comment:
I assume (hope) this code is generated - where is the generator?

Reply:
I can add it to the patch, but it will fail tests and etc. I just whipped this patch up to get this subject moving, it is not a high priority for me, I have some complex regex bugs on my TODO list. Are you just horrified at the generated code or is outright wrong?  Fwiw, the file is result of a two step process. First this:

```
git grep -P "[@%$]\{?\^." pod/perlvar.pod | perl -lne'/(?<!\S)[\@\%\$]\{?(\^.\w*)\}?/ and print "$1"' | sort | uniq > caret_var.txt
```

And then this:

```
# cat ~/scratch/gen.pl 
use strict;
use warnings;
use warnings FATAL => 'all';
open my $fh, "<", "lib/English.pm";
my %symbol;
while (<$fh>) {
    chomp;
    if (/^\s*\*(\w+)\s+=\s+\*(\^\S|\S)(?:\{(\w+)\})?\s+;/) {
        my ( $long, $short, $type ) = ( $1, $2, $3 );
        $type = "ALL" unless $type;
        my $fc = ord( substr( $long, 0, 1 ) ) - 64;
        $symbol{ length $long }{$fc}{$long} =
          { short => $short, type => $type };
    }
}
open $fh, "<", "caret_var.txt";
while (<$fh>) {
    chomp;
    next unless /^\^(\S+)/;
    my $long= $1;
    my $fc = ord( substr( $long, 0, 1 ) ) - 64;
    if ($fc < 1) {
        warn "ignoring: $long";
        next;
    }
    if ($symbol{length $long}{$fc}{$long}) {
        next;
    }
    $symbol{length $long}{$fc}{$long} = { short => $long, type => "STD" };
}

foreach my $length ( sort { $a <=> $b } keys %symbol ) {
    print "case $length:\n";
    my $nkeys = 0 + keys %{ $symbol{$length} };
    print "    switch ((U8)name[0]) {\n";
    my @fc= sort { $a <=> $b } keys %{ $symbol{$length} }; 
    foreach my $fc ( @fc ) {
        print "    case $fc:\n";
        foreach my $long ( sort keys %{ $symbol{$length}{$fc} } ) {
            my $short           = $symbol{$length}{$fc}{$long}{short};
            my $type            = $symbol{$length}{$fc}{$long}{type};

            my $longname_dquoted = $long;
            $longname_dquoted =~ s/^(.)/sprintf "\\%03o",ord($1)-64/e;
            my $shortname_dquoted = $short;
            my $shortname_squoted;
            my $shortlen;

            if ( length($shortname_dquoted)>1 and $shortname_dquoted =~ s/^\^// )
            {
                if (length($shortname_dquoted) == 1) {
                    $shortlen = 1;
                    $shortname_squoted = sprintf "'\\%03o'",ord($shortname_dquoted)-64;
                    $shortname_dquoted = sprintf qq("\\%03o"),ord($shortname_dquoted)-64;
                } else {
                    $shortlen = length($shortname_dquoted);
                    $shortname_dquoted = sprintf qq("\\%03o%s"),
                      ord($shortname_dquoted) - 64, length($shortname_dquoted)>1 ? substr($shortname_dquoted,1) : "";
                }
            }
            else {
                $shortlen = 1;
                $shortname_squoted = $shortname_dquoted;
                $shortname_squoted =~ s/([\\'])/\\$1/;
                $shortname_dquoted =~ s/([\\"])/\\$1/;
                $shortname_squoted = sprintf q('%s'), $shortname_squoted;
                $shortname_dquoted = sprintf q("%s"), $shortname_dquoted;
            }
            $longname_dquoted = qq("$longname_dquoted");

            my $ok = 1;    #$short ne "!" ? "1" : "0";
            0 and print <<~"EOF";
                if ($ok) {
                    /* \${^$long} -> \$$short  ($type) */
                    SV *gv_as_sv= (SV*)gv_fetchpvs($shortname_dquoted,GV_ADDMG|GV_ADDMULTI,SVt_PVGV);
                    if (hv_stores(PL_defstash,$longname_dquoted,gv_as_sv))
                        SvREFCNT_inc(gv_as_sv);
                }
            EOF
            if ($type ne "STD") {
                my $len = length $long;
                if ($shortname_squoted) {
                    print <<~"EOF";
                            /* \${^$long} -> \$$short  ($type) */
                            if (memEQs(name,*lenp,$longname_dquoted)) {
                                tmp_buf[0] = $shortname_squoted;
                                tmp_buf[1] = 0;
                                *lenp = 1;
                                return tmp_buf;
                            }
                    EOF
                } else {
                    print <<~"EOF";
                            /* \${^$long} -> \$$short  ($type) */
                            if (memEQs(name,*lenp,$longname_dquoted)) {
                                Copy($shortname_dquoted,tmp_buf,$shortlen,char);
                                *lenp = $shortlen;
                                tmp_buf[*lenp] = 0;
                                return tmp_buf;
                            }
                    EOF
                }
            } elsif ($length > 1)  {
                print <<~"EOF";
                        /* \${^$long} */
                        if (memEQs(name,*lenp,$longname_dquoted))
                            return name;
                EOF
            } elsif ($fc == $fc[-1]) {
                print <<~"EOF";
                        /* \$^$long */
                        return name;
                EOF
            } else {
                print <<~"EOF";
                        /* \$^$long */
                        /* FALLTHROUGH */;
                EOF
            }
        }
        print "        break;\n" if $length>1;
    }
    print "    }\n";
    print "    break;\n";
}
```

Reply:
I was thinking more of what happens when someone wants to add a new name?

For the generated code it's mostly "maybe I wouldn't do it quite that way" with the advantage of having seen this generated code.

Reply:
>I was thinking more of what happens when someone wants to add a new name?

Perhaps it was naive of me, but I figured adding one name to this type of code wouldn't be too difficult.  That was partly why i structured it like this. A bit of copy and paste and you are good.

> For the generated code it's mostly "maybe I wouldn't do it quite that way" with the advantage of having seen this generated code.

Reading it again with fresh eyes, I think I would agree with you and would certainly be open to suggestions. :-)